### PR TITLE
Simplify CompositeModelBinder

### PR DESCRIPTION
**Pull request by rynowak**
Also adds a diagnostic ToString to ModelBindingResult to make testing
easier.

### CODE REVIEW

**Comment by rynowak** [comment]
This change should remove a task allocation for every model binding operation. No test changes were needed, confirming my suspicion that the scenarios covered by this were fixed elsewhere. 

Currently working on before/after perf results.

**Comment by rynowak** [comment]
Before:
image

After:
image

Removes about 15% of the allocated tasks. Not thrilling by any means, but a good savings for something that's basically code cleanup anyway.

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs [hunk 1]
```diff
@@ -130,6 +116,23 @@ private ModelBindingResult(string key, object model, bool isModelSet, ModelValid
         public ModelValidationNode ValidationNode { get; }
 
         /// <inheritdoc />
+        public override string ToString()
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs]: Is this for debugging ease?

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs [hunk 2]
```diff
@@ -130,6 +116,23 @@ private ModelBindingResult(string key, object model, bool isModelSet, ModelValid
...
+        {
+            if (Key == null)
+            {
+                return "No Result";
+            }
+            else if (IsModelSet)
+            {
+                return $"Success: {Key} -> {Model}";
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs]: Suggest single quotes around `'{Model}'`. Will make things more clear when its `ToString()` includes spaces.

**Comment by dougbu** [comment]
> No test changes were needed, confirming my suspicion that the scenarios covered by this were fixed elsewhere.

Nice. Also confirms my suspicion no code reads `Key` of the top-level `ModelBindingResult`.

**Comment by dougbu** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs]: Yes, and also without this the messages printed by xUnit when you have a test failure are pretty un-actionable.

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs]: sure

**Pull request closed without merging**